Fix plvpwaa dlc error

Rokkubro opened the pull request:
Return 1 for the number of output entries in the GetNsDataIdList1 stub. This fixes the extra content for Professor Layton vs Phoenix Wright Ace Attorney as the game expects the boss extdata to not be empty. Might break other games if they attempt to do anything with the ns data. (although the readnsdata and deletensdata methods are both still stubbed)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/6436)
<!-- Reviewable:end -->

SachinVin commented:
I don't like that this potentially breaks multiple others, in favor of fixing one game.

Rokkubro commented:
I’ve just pushed a commit that only makes this change if this specific game is running.

Potentially the alternative is to check for the number of files in the boss extdata directory but I think that makes too many assumptions about how the spotpass service works without actually doing a proper implementation.

SachinVin commented:
No game specific code in core please.

PydraxAlpta commented:
Can we figure out what games would break because of the first commit? I don't like a game specific exception clause but it seems that without knowing what would break and how it breaks this will be at a stalemate to not progress. What other games are known to try and access the ns data but are auto prevented by this line returning an empty boss extra data?

Rokkubro commented:
Ok, so now the BOSS service will attempt to check the number of files in the boss extdata directory to determine the number of output entries; if it fails or if the directory is empty it will return a 0 which should preserve the original behaviour for all other games. (and will also preserve the behaviour for PLvPWAA if the boss directory is empty.

BreadFish64 commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -196,7 +247,7 @@ void Module::Interface::GetNsDataIdList(Kernel::HLERequestContext& ctx) {
     IPC::RequestBuilder rb = rp.MakeBuilder(3, 2);
     rb.Push(RESULT_SUCCESS);
-    rb.Push<u16>(0); /// Actual number of output entries
+    rb.Push<u16>(GetOutputEntriesCount()); /// Actual number of output entries
```
This is dangerous since this can be larger than the buffer size and games might access out-of-bounds memory.
https://www.3dbrew.org/wiki/BOSSU:GetNsDataIdList
It doesn't seem too hard to implement properly, but this output at least needs to be the minimum of this and `max_entries`.

Rokkubro replied:
Ok so now there should be an array of zeroed nsdataids for the games to read in case, hopefully I used the mappedbuffer correctly here. I've done a bit of snooping with about 15 titles that use spotpass extdata and it should be relatively simple to read the actual IDs from the files each title's boss directory, and also return the payload data from each entry should a game request it. I'm not sure that's within the scope of this branch though.

Rokkubro replied:
Gotcha, with the new code it should only ever return a number of entries that is less than max_entries and has a readable nsdataid in the buffer. Also no longer using any VLAs, are std::vector and dynamic arrays allowed?

Rokkubro commented:
Made a rather large change so the service will actually read the NsDataIds from the files in the spotpass directory when GetNsDataIdList(1/2/3) is called, and will also attempt to return data from those files if ReadNsData is called.

SachinVin commented:
Hi, is this actually how the 3ds calculates GetNsDataIdList?

Rokkubro commented:
I think it’s pretty close. On the 3DS the service checks that the payload size is larger than or equal to the size in the header, here I’ve just made it equal. Also I haven’t implemented the filter parameter or distinguished between the 4 GetNsIdList functions. 

At [3dbrew](https://www.3dbrew.org/wiki/BOSSU:GetNsDataIdList) it says the filter parameter is checked against the datatype in the header and if it fails the nsdataid is not added to the list. However I haven’t been able to reproduce this at all on a real 3DS, the datatype just seems to be ignored.

SachinVin commented:
So, does the 3ds also iterate over the folders in `boss` and return the List ?

Rokkubro commented:
Yes that's right. The service actually seems to be doing some filtering of files before opening them just based on the file names. I haven't worked out exactly how that works though. 3dbrew states :"The contentdatatype is extracted from the extdata filename, which originates from the BOSS-container." so it might be that the datatype field in the header is ignored completely and it determines it from the filename somehow.

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+    const u32 low = static_cast<u32>(extdata_id & 0xFFFFFFFF);
+
+    return FileSys::ConstructExtDataBinaryPath(1, high, low);
+}
```
newline before and after this function

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+    std::vector<NsDataEntry> ns_data;
+    u32 entry_count = 0;
+    u32 files_to_read = 100;
+    FileSys::Entry boss_files[100];
```
can be an std::array
dito below

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+}
+std::vector<NsDataEntry> Module::Interface::GetNsDataEntries(u32 max_entries) {
+    std::vector<NsDataEntry> ns_data;
+    u32 entry_count = 0;
```
the variable declaration can be moved down further

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+std::vector<NsDataEntry> Module::Interface::GetNsDataEntries(u32 max_entries) {
+    std::vector<NsDataEntry> ns_data;
+    u32 entry_count = 0;
+    u32 files_to_read = 100;
```
can be const

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+    }
+
+    for (u32 i = 0; i < entry_count; i++) {
+        if (boss_files[i].is_directory || boss_files[i].file_size < 52) {
```
replace this magic number with a meaningful variable or add a comment

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+        FileSys::Path boss_path{GetBossDataDir()};
+        auto archive_result = extdata_archive_factory.OpenSpotpass(boss_path, 0);
+
+        if (archive_result.Succeeded()) {
```
```suggestion
        if (!archive_result.Succeeded()) {
            LOG_WARNING(Service_BOSS, "Extdata opening failed");
            return;
        }
```
same with the other check below
it also seems like the extdata opening can happen outside the loop.

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+                // signature in the header So the total header is only 52 bytes long
+                u8 header_length[1];
+                file->Read(0, 1, header_length);
+                if (header_length[0] != 0x18) {
```
magic number being used here.

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+                Core::System::GetInstance().GetAppLoader().ReadProgramId(program_id);
+                u8 spotpass_program_id_array[8];
+                u64 spotpass_program_id = 0;
+                file->Read(0x18, 8, spotpass_program_id_array);
```
Consider declaring the Spotpass data as a struct and reading it in one go.
that could be a lot cleaner.

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+    FileSys::Path boss_path{GetBossDataDir()};
+
+    auto archive_result = extdata_archive_factory.OpenSpotpass(boss_path, 0);
+    if (archive_result.Succeeded()) {
```
consider inverting this check and returning early on errors here

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+            entry_count = dir->Read(files_to_read, files);
+            LOG_DEBUG(Service_BOSS, "Spotpass Extdata directory contains {} files", entry_count);
+        } else {
+            LOG_WARNING(Service_BOSS, "Extdata directory opened unsuccessfully :(");
```
```suggestion
            LOG_WARNING(Service_BOSS, "Extdata directory opening failed");
```

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -185,6 +190,167 @@ void Module::Interface::GetStepIdList(Kernel::HLERequestContext& ctx) {
+    for (u32 i = 0; i < entry_count; i++) {
+        output_entries.push_back(ns_data[i].ns_data_id);
+    }
+    buffer->Write(&output_entries[0], 0, sizeof(u32) * entry_count);
```
```suggestion
    buffer->Write(output_entries.data(), 0, sizeof(u32) * entry_count);
```

FearlessTobi commented on src/core/hle/service/boss/boss.cpp:
```diff
@@ -536,10 +710,59 @@ void Module::Interface::ReadNsData(Kernel::HLERequestContext& ctx) {
+                                    size, offset, entry.payload_size);
+                        continue;
+                    }
+                    u8* ns_data_array = new u8[size];
```
do not use `new`.
can't this just be a vector?

FearlessTobi reviewed (changes requested):
Thanks for implementing this!
There are some things that need to be fixed though.

Rokkubro commented:
@FearlessTobi I've made all the changes as you requested. Thanks for the feedback! Apologies for my poor c++ 😅

Also added in filtering by datatype as was done on the 3ds, not really used in PLvPWAA but other games do (like MK7, swapdoodle).

@PydraxAlpta A bunch of games (Mario Kart 7, Swapdoodle, Smash Bros) will infinitely hang or crash with the first commit it seems as they try to request data from the phantom entry we say exists. Mario Kart 7 attempts to do this on startup and will never reach the main menu :(

Removed the ReadNsData implementation as games will call GetNsDataHeaderInfo beforehand and I haven't finished the implementation of that.

So far I haven't worked out exactly how the filenames are encoded but I have discovered 2 important things. 
1. It contains characters representing NsDataID, payload size and datatype in that order.
2. When the BOSS service writes the data out to the extdata archive, the values recorded in the filename never differ from those stored in the header.
So it should be fine to continue to use the values from the header, even though the real 3ds gets the datatype parameter from the filename.

I think this is as far as I want to go for this pull request, I've made a new branch on my citra clone where I will work on implementing the rest of the spotpass service. For now PLvPWAA will work as long as valid spotpass data is stored in its boss extdata directory, and all other games will work as before so long as their boss extdata directory is empty.

Rokkubro commented:
Superseded by #6549
